### PR TITLE
Add `String.is_valid_unicode_identifier()`

### PR DESCRIPTION
**Pull request by timothyqiu**
Implements https://github.com/godotengine/godot-proposals/discussions/9851

Rationale:

- The old `is_valid_identifier()` is checking for traditional ASCII identifiers. Its name can be confusing because GDScript implements Unicode identifiers.
- There is currently no way to check according to GDScript's rule of Unicode identifiers. `TextServer.is_valid_identifier()` is close, but advanced text server adds extra conversions and rules. Text server extensions could potentially also change the rules.

This PR:

- Adds `is_valid_unicode_identifier()`: Basic unicode identifier checking, the same as how GDScript parses identifiers
- Adds `is_valid_ascii_identifier()` : Same as the old `is_valid_identifier()`
- Deprecates `is_valid_identifier()`
- Renames internal method `validate_identifier()` to `validate_ascii_identifier()`

### CODE REVIEW

**Review by Calinou — APPROVED** [review]
This change makes sense to me. I've been wondering how to check for a valid Unicode identifier recently from a script, in fact :slightly_smiling_face:

File: doc/classes/String.xml [hunk 1]
```diff
@@ -452,6 +452,19 @@
 				Returns [code]true[/code] if all characters of this string can be found in [param text] in their original order, [b]ignoring case[/b].
 			</description>
 		</method>
+		<method name="is_valid_ascii_identifier" qualifiers="const">
+			<return type="bool" />
+			<description>
+				Returns [code]true[/code] if this string is a valid ASCII identifier. A valid ASCII identifier may contain only letters, digits and underscores ([code]_[/code]), and the first character may not be a digit.
```
Comment (AThousandShips) [inline doc/classes/String.xml:458]:
```suggestion
				Returns [code]true[/code] if this string is a valid ASCII identifier. A valid ASCII identifier may contain only letters, digits, and underscores ([code]_[/code]), and the first character may not be a digit.
```

File: doc/classes/StringName.xml [hunk 2]
```diff
@@ -421,6 +421,19 @@
 				Returns [code]true[/code] if all characters of this string can be found in [param text] in their original order, [b]ignoring case[/b].
 			</description>
 		</method>
+		<method name="is_valid_ascii_identifier" qualifiers="const">
+			<return type="bool" />
+			<description>
+				Returns [code]true[/code] if this string is a valid ASCII identifier. A valid ASCII identifier may contain only letters, digits and underscores ([code]_[/code]), and the first character may not be a digit.
```
Comment (AThousandShips) [inline doc/classes/StringName.xml:427]:
```suggestion
				Returns [code]true[/code] if this string is a valid ASCII identifier. A valid ASCII identifier may contain only letters, digits, and underscores ([code]_[/code]), and the first character may not be a digit.
```

**Review by AThousandShips — APPROVED** [review]

File: doc/classes/String.xml [hunk 3]
```diff
@@ -522,6 +535,21 @@
...
+		<method name="is_valid_unicode_identifier" qualifiers="const">
+			<return type="bool" />
+			<description>
+				Returns [code]true[/code] if this string is a valid Unicode identifier.
+				[codeblock]
+				print("node_2d".is_valid_unicode_identifier())    # Prints true
+				print("TYPE_FLOAT".is_valid_unicode_identifier()) # Prints true
+				print("1st_method".is_valid_unicode_identifier()) # Prints false
+				print("MyMethod#2".is_valid_unicode_identifier()) # Prints false
+				print("体力".is_valid_unicode_identifier())       # Prints true
```
Comment (Mickeon) [inline doc/classes/String.xml:547]:
```suggestion
				print("体力".is_valid_unicode_identifier())        # Prints true
```

File: doc/classes/StringName.xml [hunk 4]
```diff
@@ -491,6 +504,21 @@
...
+		<method name="is_valid_unicode_identifier" qualifiers="const">
+			<return type="bool" />
+			<description>
+				Returns [code]true[/code] if this string is a valid Unicode identifier.
+				[codeblock]
+				print("node_2d".is_valid_unicode_identifier())    # Prints true
+				print("TYPE_FLOAT".is_valid_unicode_identifier()) # Prints true
+				print("1st_method".is_valid_unicode_identifier()) # Prints false
+				print("MyMethod#2".is_valid_unicode_identifier()) # Prints false
+				print("体力".is_valid_unicode_identifier())       # Prints true
```
Comment (Mickeon) [inline doc/classes/StringName.xml:516]:
```suggestion
				print("体力".is_valid_unicode_identifier())        # Prints true
```

File: doc/classes/String.xml [hunk 5]
```diff
@@ -522,6 +535,21 @@
...
+				[/codeblock]
+				See also [method is_valid_ascii_identifier].
+				[b]Note:[/b] This method performs basic Unicode identifier checking, which is the same as how GDScript parses identifiers. See [method TextServer.is_valid_identifier] for detailed rules and potential advanced checks that utilize other Unicode features.
```
Comment (Mickeon) [inline doc/classes/String.xml:550]: Nitpick. I feel like this note can be shortened. I also get you're trying to direct readers interested in the "advanced checks", but it's a lot of alienating 
```suggestion
				[b]Note:[/b] This method performs basic Unicode identifier checking, which is the same as how GDScript parses identifiers. See [method TextServer.is_valid_identifier] for more details.
```
Or:
```suggestion
				[b]Note:[/b] This method checks identifiers the same way as GDScript. See [method TextServer.is_valid_identifier] for more advanced checks.
```
Or somewhat similar.

File: doc/classes/StringName.xml [hunk 6]
```diff
@@ -491,6 +504,21 @@
...
+				[/codeblock]
+				See also [method is_valid_ascii_identifier].
+				[b]Note:[/b] This method performs basic Unicode identifier checking, which is the same as how GDScript parses identifiers. See [method TextServer.is_valid_identifier] for detailed rules and potential advanced checks that utilize other Unicode features.
```
Comment (Mickeon) [inline doc/classes/StringName.xml:519]: Same comment as above

**Review by Mickeon — COMMENTED** [review]
The first line of the new method's description could be a bit more detailed to entail what a "valid Unicode identifier" means, at least roughly.

The examples could also be more distinct than `is_valid_identifier`'s, such as using accented letters and similar.

File: doc/classes/String.xml (hunk 3 above)
Comment (timothyqiu) [reply doc/classes/String.xml:547]: This depends on the font used for rendering. Usually, when using a monospace font, each Chinese character is two Latin characters wide. GitHub might show them as not aligned because it's using different fonts for them.

**Review by Mickeon — APPROVED** [review]

File: scene/gui/tree.cpp [hunk 7]
```diff
@@ -4495,9 +4495,16 @@ void Tree::item_edited(int p_column, TreeItem *p_item, MouseButton p_custom_mous
 }
 
 void Tree::item_changed(int p_column, TreeItem *p_item) {
-	if (p_item != nullptr && p_column >= 0 && p_column < p_item->cells.size()) {
-		p_item->cells.write[p_column].dirty = true;
-		columns.write[p_column].cached_minimum_width_dirty = true;
+	if (p_item != nullptr) {
```
Comment (bruvzg) [inline scene/gui/tree.cpp:4498]: Tree change seems unrelated to identifiers (change itself seems ok, but I'm when it's necessary).

**Review by bruvzg — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!
